go/vt/vtctl/workflow/server.go

```diff
@@ -419,7 +420,72 @@ func (s *Server) GetWorkflows(ctx context.Context, req *vtctldatapb.GetWorkflows
 		return nil, err
 	}
 
-	m := sync.Mutex{} // guards access to the following maps during concurrent calls to scanWorkflow
+	m := sync.Mutex{} // guards access to the following maps during concurrent calls to fetchCopyStates and scanWorkflow
+
+	copyStatesByShardStreamId := make(map[string][]*vtctldatapb.Workflow_Stream_CopyState, len(results))
+
+	fetchCopyStates := func(ctx context.Context, tablet *topo.TabletInfo, streamIds []int64) error {
+		span, ctx := trace.NewSpan(ctx, "workflow.Server.fetchCopyStates")
+		defer span.Finish()
+
+		span.Annotate("keyspace", req.Keyspace)
+		span.Annotate("shard", tablet.Shard)
+		span.Annotate("tablet_alias", tablet.AliasString())
+
+		copyStates, err := s.getWorkflowCopyStates(ctx, tablet, streamIds)
+		if err != nil {
+			return err
+		}
+
+		m.Lock()
+		defer m.Unlock()
+
+		for _, copyState := range copyStates {
+			shardStreamId := fmt.Sprintf("%s/%d", tablet.Shard, copyState.StreamId)
+			copyStatesByShardStreamId[shardStreamId] = append(
+				copyStatesByShardStreamId[shardStreamId],
+				copyState,
+			)
+		}
+
+		return nil
+	}
+
+	var (
+		fetchCopyStatesWg     sync.WaitGroup
+		fetchCopyStatesErrors concurrency.FirstErrorRecorder
```

Comment:
I apologize if we discussed this before, but this seems almost tailor made for `sync.errgroup`. Any reason(s) not to leverage that? It can cut the work short when there's an error too (`errgroup.WithContext`) since it's wasted effort after that in this case, and that could be a lot of effort when we're talking about 10s or even 100s of thousands of shard streams.

You can see examples in the files being changed in the PR.

Reply:
Hm there are no examples in the files changed in this PR of `errgroup` usage.

```sh
$ grep errgroup $(git diff --name-only main | xargs)
$
```

However, I'm familiar with the API and happy to rework this code to use it if you prefer.

Reply:
```
❯ git grep errgroup | grep -E 'wrangler|workflow'
go/vt/vtctl/workflow/traffic_switcher.go:	"golang.org/x/sync/errgroup"
go/vt/vtctl/workflow/traffic_switcher.go:	searchGroup, gctx := errgroup.WithContext(ctx)
go/vt/vtctl/workflow/traffic_switcher.go:	initGroup, gctx := errgroup.WithContext(ctx)
go/vt/wrangler/traffic_switcher.go:	"golang.org/x/sync/errgroup"
go/vt/wrangler/traffic_switcher.go:	searchGroup, gctx := errgroup.WithContext(ctx)
go/vt/wrangler/traffic_switcher.go:	initGroup, gctx := errgroup.WithContext(ctx)
```